File: sdk/javascript/src/sync.integration.test.js
const test = require('node:test')
const assert = require('node:assert/strict')
const crypto = require('node:crypto')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')
const { execFileSync } = require('node:child_process')
const { unzipSync, strFromU8 } = require('fflate')

const BASE_URL = (process.env.VOLA_TEST_URL || '').replace(/\/+$/, '')
const DEV_SLUG_CANDIDATES = (process.env.VOLA_TEST_DEV_SLUGS || 'demo,de,admin')
  .split(',')
  .map((value) => value.trim())
  .filter(Boolean)
const ROOT = path.resolve(__dirname, '../../..')
const FIXTURE_DIR = path.join(ROOT, 'internal', 'services', 'testdata')

function runVola(args) {
  const configured = process.env.VOLA_CLI
  if (configured) {
    execFileSync(configured, args, { stdio: 'inherit' })
    return
  }
  if (fs.existsSync('/tmp/vola')) {
    execFileSync('/tmp/vola', args, { stdio: 'inherit' })
    return
  }
  execFileSync('go', ['run', './cmd/vola', ...args], { cwd: ROOT, stdio: 'inherit' })
}

function skipIfNoServer() {
  if (!BASE_URL) {
    return true
  }
  return false
}

function loadSkillFixture() {
  const archive = unzipSync(new Uint8Array(fs.readFileSync(path.join(FIXTURE_DIR, 'ahub-sync.skill'))))
  const files = {}
  for (const [name, data] of Object.entries(archive)) {
    if (!name.startsWith('pkg-skill/') || name.endsWith('/')) continue
    files[name.replace(/^pkg-skill\//, '')] = strFromU8(data)
  }
  return files
}

function loadPlan() {
  return JSON.parse(fs.readFileSync(path.join(FIXTURE_DIR, 'sync-fixture-plan.json'), 'utf8'))
}

function expandedBinary(base, seed, multiplier) {
  const targetSize = base.length + Math.max(multiplier, 1) * (256 << 10)
  const chunks = [base]
  let size = base.length
  let counter = 0
  while (size < targetSize) {
    const chunk = crypto.createHash('sha256').update(`${seed}:${counter}`).digest()
    chunks.push(chunk)
    size += chunk.length
    counter += 1
  }
  return Buffer.concat(chunks).subarray(0, targetSize)
}

function materializeSource(multiplier = 1) {
  const tmpRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'vola-js-sync-'))
  const files = loadSkillFixture()
  const plan = loadPlan()
  const binary = fs.readFileSync(path.join(FIXTURE_DIR, 'tiny.png'))

  for (const skillName of plan.skill_names) {
    const skillRoot = path.join(tmpRoot, skillName)
    fs.mkdirSync(skillRoot, { recursive: true })
    for (const [relPath, content] of Object.entries(files)) {
      const target = path.join(skillRoot, relPath)
      fs.mkdirSync(path.dirname(target), { recursive: true })
      fs.writeFileSync(target, content, 'utf8')
    }
    for (const extra of plan.extra_text_files) {
      const target = path.join(skillRoot, extra.path)
      fs.mkdirSync(path.dirname(target), { recursive: true })
      fs.writeFileSync(target, (files[extra.source] + '\n').repeat(extra.repeat * Math.max(multiplier, 1)), 'utf8')
    }
    for (const relPath of plan.binary_assignments[skillName] || []) {
      const target = path.join(skillRoot, relPath)
      fs.mkdirSync(path.dirname(target), { recursive: true })
      fs.writeFileSync(target, expandedBinary(binary, `${skillName}:${relPath}`, multiplier))
    }
  }
  return tmpRoot
}

async function registerUser() {
  const slug = `js-sync-${Date.now()}`
  const email = `${slug}@test.local`
  const password = 'vola-sync-1234'
  const registerRes = await fetch(`${BASE_URL}/api/auth/register`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ slug, email, password }),
  })
  let jwtToken
  if (registerRes.status !== 404) {
    assert.equal(registerRes.status, 201)
    const registerBody = await registerRes.json()
    jwtToken = registerBody.access_token
  } else {
    let devTokenBody
    let lastStatus = 0
    for (const candidate of DEV_SLUG_CANDIDATES) {
      const devTokenRes = await fetch(`${BASE_URL}/api/auth/token/dev`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ slug: candidate }),
      })
      lastStatus = devTokenRes.status
      if (devTokenRes.status === 404) continue
      assert.equal(devTokenRes.status, 200)
      devTokenBody = await devTokenRes.json()
      break
    }
    assert.ok(devTokenBody, `expected one of ${DEV_SLUG_CANDIDATES.join(', ')} to exist, last status ${lastStatus}`)
    jwtToken = devTokenBody.token
  }

  const tokenRes = await fetch(`${BASE_URL}/api/tokens`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${jwtToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      name: 'js-sync-test',
      scopes: ['read:bundle', 'write:bundle'],
      max_trust_level: 3,
      expires_in_days: 1,
    }),
  })
  assert.equal(tokenRes.status, 201)
  const tokenBody = await tokenRes.json()
  return tokenBody.data ? tokenBody.data.token : tokenBody.token
}

function readManifestFromArchive(archivePath) {
  const files = unzipSync(new Uint8Array(fs.readFileSync(archivePath)))
  return JSON.parse(strFromU8(files['manifest.json']))
}

test('Vola JS SDK handles json bundle and archive sessions', { skip: skipIfNoServer() }, async () => {
  const { Vola } = require('../dist/index.js')
  const token = await registerUser()
  const sourceDir = materializeSource(2)
  const bundlePath = path.join(os.tmpdir(), `vola-js-${Date.now()}.ndrv`)
  const archivePath = path.join(os.tmpdir(), `vola-js-${Date.now()}.ndrvz`)

  runVola(['sync', 'export', '--source', sourceDir, '-o', bundlePath])
  runVola(['sync', 'export', '--source', sourceDir, '--format', 'archive', '-o', archivePath])

  const bundle = JSON.parse(fs.readFileSync(bundlePath, 'utf8'))
  const manifest = readManifestFromArchive(archivePath)
  const archiveBytes = new Uint8Array(fs.readFileSync(archivePath))
  const hub = new Vola({ baseURL: BASE_URL, token })

  const authInfo = await hub.getAuthInfo()
  assert.equal(authInfo.api_base, BASE_URL)
  assert.deepEqual(authInfo.scopes, ['read:bundle', 'write:bundle'])

  const preview = await hub.previewBundle(bundle)
  assert.ok(preview.fingerprint)

  const imported = await hub.importBundle(bundle)
  assert.ok(imported.files_written > 0)

  const exported = await hub.exportBundle('json')
  assert.ok(exported.skills)

  const session = await hub.startSyncSession({
    transport_version: 'ahub.sync/v1',
    format: 'archive',
    mode: 'merge',
    manifest,
    archive_size_bytes: archiveBytes.length,
    archive_sha256: manifest.archive_sha256,
  })
  const resumed = await hub.resumeSession(session.session_id, archiveBytes)
  assert.ok(['ready', 'uploading'].includes(resumed.status))
  const committed = await hub.commitSession(session.session_id)
  assert.ok(committed.files_written > 0)

  const jobs = await hub.listSyncJobs()
  assert.ok(jobs.length >= 3)
})

File: internal/cli/sync_integration_test.go
package cli

import (
	"archive/zip"
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/agi-bar/vola/internal/models"
	"github.com/agi-bar/vola/internal/runtimecfg"
	"github.com/agi-bar/vola/internal/services"
)

var (
	volaBinaryOnce sync.Once
	volaBinaryPath string
	volaBinaryErr  error
)

type syncFixturePlan struct {
	SkillNames        []string              `json:"skill_names"`
	ExtraTextFiles    []syncFixtureTextFile `json:"extra_text_files"`
	BinaryAssignments map[string][]string   `json:"binary_assignments"`
}

type syncFixtureTextFile struct {
	Path   string `json:"path"`
	Source string `json:"source"`
	Repeat int    `json:"repeat"`
}

type cliSessionEnvelope struct {
	OK   bool                       `json:"ok"`
	Data models.SyncSessionResponse `json:"data"`
}

func TestAgenthubSyncLocalSQLiteRoundTrip_WithRealisticFixture(t *testing.T) {
	binary := buildAgenthubBinary(t)
	env, _, _, _, workDir := isolatedAgenthubEnv(t)
	sourceDir := materializeFixtureSource(t, 2)
	archivePath := filepath.Join(workDir, "fixture.ndrvz")
	pulledPath := filepath.Join(workDir, "fixture-pulled.ndrvz")

	mustRunAgenthub(t, binary, env, "sync", "export", "--source", sourceDir, "--format", "archive", "-o", archivePath)

	stdout, _ := mustRunAgenthub(t, binary, env, "sync", "preview", "--bundle", archivePath)
	if !strings.Contains(stdout, "\"fingerprint\"") {
		t.Fatalf("preview output missing fingerprint: %s", stdout)
	}

	mustRunAgenthub(t, binary, env, "sync", "push", "--bundle", archivePath, "--transport", "archive")
	mustRunAgenthub(t, binary, env, "sync", "pull", "--format", "archive", "-o", pulledPath)

	stdout, stderr, code := runAgenthub(t, binary, env, "sync", "diff", "--left", archivePath, "--right", pulledPath)
	if code != 0 {
		t.Fatalf("diff exit code = %d\nstdout:\n%s\nstderr:\n%s", code, stdout, stderr)
	}
	if !strings.Contains(stdout, "Equal: yes") {
		t.Fatalf("diff output missing equality marker: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "sync", "history")
	if !strings.Contains(stdout, "\"direction\": \"import\"") || !strings.Contains(stdout, "\"direction\": \"export\"") {
		t.Fatalf("history output missing import/export jobs: %s", stdout)
	}
	mustRunAgenthub(t, binary, env, "daemon", "stop")
}

func TestAgenthubSyncResume_LocalSQLiteArchiveSession(t *testing.T) {
	binary := buildAgenthubBinary(t)
	env, configPath, statePath, _, workDir := isolatedAgenthubEnv(t)
	sourceDir := materializeFixtureSource(t, 12)
	archivePath := filepath.Join(workDir, "large.ndrvz")
	pulledPath := filepath.Join(workDir, "large-pulled.ndrvz")

	mustRunAgenthub(t, binary, env, "sync", "export", "--source", sourceDir, "--format", "archive", "-o", archivePath)
	mustRunAgenthub(t, binary, env, "sync", "history")

	cfg := loadCLIConfigForTest(t, configPath)
	state := loadRuntimeStateForTest(t, statePath)
	if strings.TrimSpace(cfg.Local.OwnerToken) == "" {
		t.Fatal("expected local owner token after history bootstrap")
	}

	archiveBytes, err := os.ReadFile(archivePath)
	if err != nil {
		t.Fatalf("read archive: %v", err)
	}
	_, manifest, err := services.ParseBundleArchive(archiveBytes)
	if err != nil {
		t.Fatalf("parse archive: %v", err)
	}
	started := startSyncSessionForTest(t, state.APIBase, cfg.Local.OwnerToken, archiveBytes, manifest)
	partSize := int(started.ChunkSizeBytes)
	if partSize <= 0 {
		t.Fatalf("invalid chunk size: %d", started.ChunkSizeBytes)
	}
	uploadSyncPartForTest(t, state.APIBase, cfg.Local.OwnerToken, started.SessionID.String(), 0, archiveBytes[:partSize])

	sessionFile := archivePath + ".session.json"
	sessionState := map[string]any{
		"api_base":            state.APIBase,
		"bundle_path":         archivePath,
		"session_id":          started.SessionID.String(),
		"preview_fingerprint": "",
		"profile":             "",
		"created_at":          time.Now().UTC().Format(time.RFC3339),
	}
	data, err := json.MarshalIndent(sessionState, "", "  ")
	if err != nil {
		t.Fatalf("marshal session state: %v", err)
	}
	if err := os.WriteFile(sessionFile, append(data, '\n'), 0o644); err != nil {
		t.Fatalf("write session state: %v", err)
	}

	mustRunAgenthub(t, binary, env, "sync", "resume", "--bundle", archivePath)
	mustRunAgenthub(t, binary, env, "sync", "pull", "--format", "archive", "-o", pulledPath)

	stdout, stderr, code := runAgenthub(t, binary, env, "sync", "diff", "--left", archivePath, "--right", pulledPath)
	if code != 0 {
		t.Fatalf("diff exit code = %d\nstdout:\n%s\nstderr:\n%s", code, stdout, stderr)
	}
	if !strings.Contains(stdout, "Equal: yes") {
		t.Fatalf("diff output missing equality marker: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "sync", "history")
	if !strings.Contains(stdout, "\"transport\": \"archive\"") {
		t.Fatalf("history output missing archive jobs: %s", stdout)
	}
	mustRunAgenthub(t, binary, env, "daemon", "stop")
}

func TestAgenthubHostedCommands_LocalSQLiteProfile(t *testing.T) {
	binary := buildAgenthubBinary(t)
	env, configPath, statePath, _, _ := isolatedAgenthubEnv(t)

	mustRunAgenthub(t, binary, env, "sync", "history")
	cfg := loadCLIConfigForTest(t, configPath)
	state := loadRuntimeStateForTest(t, statePath)
	if strings.TrimSpace(cfg.Local.OwnerToken) == "" {
		t.Fatal("expected local owner token after bootstrap")
	}

	stdout, _ := mustRunAgenthub(t, binary, env,
		"login",
		"--profile", "official",
		"--api-base", state.APIBase,
		"--token", cfg.Local.OwnerToken,
	)
	if !strings.Contains(stdout, "Logged in to") || !strings.Contains(stdout, "Current target: profile:official") {
		t.Fatalf("unexpected login output: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "profiles")
	if !strings.Contains(stdout, "Current target: profile:official") || !strings.Contains(stdout, "* official") || !strings.Contains(stdout, state.APIBase) {
		t.Fatalf("unexpected profiles output: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "whoami")
	if !strings.Contains(stdout, "Current target: profile:official") || !strings.Contains(stdout, "Current profile: official") || !strings.Contains(stdout, "Auth mode: scoped_token") {
		t.Fatalf("unexpected whoami output: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "use", "local")
	if !strings.Contains(stdout, "Current target: local") {
		t.Fatalf("unexpected use local output: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "use", "official")
	if !strings.Contains(stdout, "Current target: profile:official") {
		t.Fatalf("unexpected use official output: %s", stdout)
	}

	stdout, _ = mustRunAgenthub(t, binary, env, "logout", "--profile", "official")
	if !strings.Contains(stdout, "Logged out profile official") || !strings.Contains(stdout, "Current target: local") {
		t.Fatalf("unexpected logout output: %s", stdout)
	}

	updated := loadCLIConfigForTest(t, configPath)
	if updated.Profiles["official"].Token != "" {
		t.Fatal("expected logout to clear saved token")
	}
	if runtimecfg.SelectedTarget(updated) != runtimecfg.TargetLocal {
		t.Fatalf("expected current target to fall back to local, got %q", runtimecfg.SelectedTarget(updated))
	}
	mustRunAgenthub(t, binary, env, "daemon", "stop")
}

func buildAgenthubBinary(t *testing.T) string {
	t.Helper()
	requireCLIIntegration(t)
	volaBinaryOnce.Do(func() {
		root := repoRoot(t)
		binDir, err := os.MkdirTemp("", "vola-cli-bin-")
		if err != nil {
			volaBinaryErr = err
			return
		}
		volaBinaryPath = filepath.Join(binDir, "vola")
		cmd := exec.Command("go", "build", "-o", volaBinaryPath, "./cmd/vola")
		cmd.Dir = root
		output, err := cmd.CombinedOutput()
		if err != nil {
			volaBinaryErr = fmt.Errorf("go build failed: %w\n%s", err, string(output))
			return
		}
	})
	if volaBinaryErr != nil {
		t.Fatal(volaBinaryErr)
	}
	return volaBinaryPath
}

func repoRoot(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	return filepath.Clean(filepath.Join(filepath.Dir(file), "..", ".."))
}

func isolatedAgenthubEnv(t *testing.T) ([]string, string, string, string, string) {
	t.Helper()
	root := t.TempDir()
	home := filepath.Join(root, "home")
	configHome := filepath.Join(root, "config")
	stateHome := filepath.Join(root, "state")
	dataHome := filepath.Join(root, "data")
	workDir := filepath.Join(root, "work")
	for _, dir := range []string{home, configHome, stateHome, dataHome, workDir} {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			t.Fatalf("mkdir %s: %v", dir, err)
		}
	}
	env := append([]string{}, os.Environ()...)
	configPath := filepath.Join(configHome, "vola", "config.json")
	statePath := filepath.Join(configHome, "vola", "runtime.json")
	env = appendOrReplaceEnv(env, "HOME", home)
	env = appendOrReplaceEnv(env, "XDG_CONFIG_HOME", configHome)
	env = appendOrReplaceEnv(env, "XDG_STATE_HOME", stateHome)
	env = appendOrReplaceEnv(env, "XDG_DATA_HOME", dataHome)
	env = appendOrReplaceEnv(env, "VOLA_CONFIG", configPath)
	for _, key := range []string{
		"VOLA_TOKEN",
		"VOLA_SYNC_TOKEN",
		"VOLA_API_BASE",
		"VOLA_SYNC_API_BASE",
		"VOLA_SYNC_PROFILE",
	} {
		env = appendOrReplaceEnv(env, key, "")
	}
	sqlitePath := filepath.Join(dataHome, "vola", "local.db")
	return env, configPath, statePath, sqlitePath, workDir
}

func runAgenthub(t *testing.T, binary string, env []string, args ...string) (string, string, int) {
	t.Helper()
	cmd := exec.Command(binary, args...)
	cmd.Env = env
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	err := cmd.Run()
	if err == nil {
		return stdout.String(), stderr.String(), 0
	}
	var exitErr *exec.ExitError
	if errors.As(err, &exitErr) {
		return stdout.String(), stderr.String(), exitErr.ExitCode()
	}
	t.Fatalf("run %v: %v", args, err)
	return "", "", 1
}

func mustRunAgenthub(t *testing.T, binary string, env []string, args ...string) (string, string) {
	t.Helper()
	stdout, stderr, code := runAgenthub(t, binary, env, args...)
	if code != 0 {
		t.Fatalf("vola %v exit=%d\nstdout:\n%s\nstderr:\n%s", args, code, stdout, stderr)
	}
	return stdout, stderr
}

func loadCLIConfigForTest(t *testing.T, path string) *runtimecfg.CLIConfig {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read config %s: %v", path, err)
	}
	var cfg runtimecfg.CLIConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		t.Fatalf("decode config %s: %v", path, err)
	}
	return &cfg
}

func loadRuntimeStateForTest(t *testing.T, path string) *runtimecfg.RuntimeState {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read runtime state %s: %v", path, err)
	}
	var state runtimecfg.RuntimeState
	if err := json.Unmarshal(data, &state); err != nil {
		t.Fatalf("decode runtime state %s: %v", path, err)
	}
	return &state
}

func startSyncSessionForTest(t *testing.T, apiBase, token string, archive []byte, manifest *models.BundleArchiveManifest) models.SyncSessionResponse {
	t.Helper()
	reqBody, err := json.Marshal(models.SyncStartSessionRequest{
		TransportVersion: models.SyncTransportVersionV1,
		Format:           models.BundleFormatArchive,
		Mode:             manifest.Mode,
		Manifest:         *manifest,
		ArchiveSizeBytes: int64(len(archive)),
		ArchiveSHA256:    manifest.ArchiveSHA256,
	})
	if err != nil {
		t.Fatalf("marshal start session: %v", err)
	}
	req, err := http.NewRequest(http.MethodPost, strings.TrimRight(apiBase, "/")+"/agent/import/session", bytes.NewReader(reqBody))
	if err != nil {
		t.Fatalf("new start session request: %v", err)
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatalf("start session request: %v", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		t.Fatalf("read start session response: %v", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		t.Fatalf("start session status=%s body=%s", resp.Status, string(body))
	}
	var envelope cliSessionEnvelope
	if err := json.Unmarshal(body, &envelope); err != nil {
		t.Fatalf("decode start session response: %v", err)
	}
	return envelope.Data
}

func uploadSyncPartForTest(t *testing.T, apiBase, token, sessionID string, index int, payload []byte) {
	t.Helper()
	req, err := http.NewRequest(http.MethodPut, strings.TrimRight(apiBase, "/")+"/agent/import/session/"+sessionID+"/parts/"+strconv.Itoa(index), bytes.NewReader(payload))
	if err != nil {
		t.Fatalf("new upload part request: %v", err)
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/octet-stream")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatalf("upload part request: %v", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		t.Fatalf("read upload part response: %v", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		t.Fatalf("upload part status=%s body=%s", resp.Status, string(body))
	}
}

func loadRealisticSkillFixture(t *testing.T) map[string]string {
	t.Helper()
	reader, err := zip.OpenReader(filepath.Join(repoRoot(t), "internal", "services", "testdata", "ahub-sync.skill"))
	if err != nil {
		t.Fatalf("open skill fixture: %v", err)
	}
	defer reader.Close()
	files := make(map[string]string)
	for _, file := range reader.File {
		if file.FileInfo().IsDir() {
			continue
		}
		relPath := strings.TrimPrefix(file.Name, "pkg-skill/")
		if relPath == file.Name || relPath == "" {
			continue
		}
		rc, err := file.Open()
		if err != nil {
			t.Fatalf("open fixture entry %s: %v", file.Name, err)
		}
		data, err := io.ReadAll(rc)
		rc.Close()
		if err != nil {
			t.Fatalf("read fixture entry %s: %v", file.Name, err)
		}
		files[relPath] = string(data)
	}
	return files
}

func readRealisticBinaryFixture(t *testing.T) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(repoRoot(t), "internal", "services", "testdata", "tiny.png"))
	if err != nil {
		t.Fatalf("read binary fixture: %v", err)
	}
	return data
}

func loadSyncFixturePlan(t *testing.T) syncFixturePlan {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(repoRoot(t), "internal", "services", "testdata", "sync-fixture-plan.json"))
	if err != nil {
		t.Fatalf("read sync fixture plan: %v", err)
	}
	var plan syncFixturePlan
	if err := json.Unmarshal(data, &plan); err != nil {
		t.Fatalf("decode sync fixture plan: %v", err)
	}
	return plan
}

func materializeFixtureSource(t *testing.T, multiplier int) string {
	t.Helper()
	if multiplier <= 0 {
		multiplier = 1
	}
	baseFiles := loadRealisticSkillFixture(t)
	binary := readRealisticBinaryFixture(t)
	plan := loadSyncFixturePlan(t)
	root := t.TempDir()
	for _, skillName := range plan.SkillNames {
		skillRoot := filepath.Join(root, skillName)
		if err := os.MkdirAll(skillRoot, 0o755); err != nil {
			t.Fatalf("create skill root: %v", err)
		}
		for relPath, content := range baseFiles {
			target := filepath.Join(skillRoot, filepath.FromSlash(relPath))
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				t.Fatalf("mkdir parent: %v", err)
			}
			if err := os.WriteFile(target, []byte(content), 0o644); err != nil {
				t.Fatalf("write base file: %v", err)
			}
		}
		for _, extra := range plan.ExtraTextFiles {
			sourceContent := baseFiles[extra.Source]
			target := filepath.Join(skillRoot, filepath.FromSlash(extra.Path))
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				t.Fatalf("mkdir extra parent: %v", err)
			}
			payload := strings.Repeat(sourceContent+"\n", extra.Repeat*multiplier)
			if err := os.WriteFile(target, []byte(payload), 0o644); err != nil {
				t.Fatalf("write extra file: %v", err)
			}
		}
		for _, relPath := range plan.BinaryAssignments[skillName] {
			target := filepath.Join(skillRoot, filepath.FromSlash(relPath))
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				t.Fatalf("mkdir binary parent: %v", err)
			}
			if err := os.WriteFile(target, expandedBinaryFixture(binary, skillName+":"+relPath, multiplier), 0o644); err != nil {
				t.Fatalf("write binary file: %v", err)
			}
		}
	}
	return root
}

func expandedBinaryFixture(base []byte, seed string, multiplier int) []byte {
	if multiplier <= 0 {
		multiplier = 1
	}
	targetSize := len(base) + multiplier*(256<<10)
	payload := make([]byte, 0, targetSize)
	payload = append(payload, base...)
	counter := 0
	for len(payload) < targetSize {
		sum := sha256.Sum256([]byte(seed + ":" + strconv.Itoa(counter)))
		payload = append(payload, sum[:]...)
		counter++
	}
	return payload[:targetSize]
}
